restructuring

adamdanielsdesign opened the pull request:
Spine "src" Restructuring issue https://github.com/pixijs/pixi-spine/issues/7

adamdanielsdesign commented:
TrackEntry is still a Class however. So for OOP reasons, not sure?
I'd say keep it like it is, actually it all looks like Esoteric actually had it broke down in their actual source originally, so basically all it's doing is sort of breaking it down to the source code, more or less.

adamdanielsdesign commented:
I say that because they formatted it with newline characters between each of the classes. If you look at the original/currect SpineRuntime, there's a newline between every class. Some between properties however.

ivanpopelyshev commented:
OOP reasons, you mean like in Java? Java at least have subclasses and static subclasses.
Scala, groovy and dart contain multiple classes in one file with no problem.

Also I dont see why we need "spine." prefix.

adamdanielsdesign commented:
OOP that TrackEntry.next, TrackEntry.animation, TrackEntry.timeScale, etc. are as Static properties.

***Maybe*** later they get used by another class.

"spine." wasn't sure if it's needed, could totally remove. I left it because it's how it's there currently.
So removing it's cool/ok though.

ivanpopelyshev commented:
OMG we have 42 new files in pull request!!!

adamdanielsdesign commented:
lol

ivanpopelyshev commented:
Anyway, give me an hour please, I'm going to checkout your gulp file and separate files that way we wont have +40 files, maximum +10 :) I'll create pull request then and we discuss it

adamdanielsdesign commented:
Sure, not in a rush.

**Note:**
 All those files are the actual classes of the SpineRuntime, I didn't go and separate out the methods & properties like I said in original post.

ivanpopelyshev commented:
Ok, I need to read more about gulp and browserify before I create my pull request.

I call on you, mighty @pkrumins !

englercj commented:
@ivanpopelyshev If you need a starting point: https://github.com/englercj/generator-browserify

adamdanielsdesign commented:
@ivanpopelyshev Should be:
```
#install
sudo npm install -g gulp
#then...
npm install
#then...
gulp
#^ would run it(the default task) same as grunt
#I have ruby installed, not sure if anything is using it though
```

adamdanielsdesign commented:
Also,
```
//bootstrap
var gulp = require('gulp'),
    concat = require('gulp-concat');
//register a 'task'
gulp.task('default',function(){
    // return an object
    // load some files
    return gulp.src(['./src/header-SpineRuntime.js','./src/SpineRuntime/*.js']) 
            //pass those files to concat, passing the -o name of the file to use
            .pipe(concat('SpineRuntime.js')) 
            // then set a destination for the -o file
            .pipe(gulp.dest('./')); 
});

```

adamdanielsdesign commented:
**Update:** Working on implementing @englercj template into the repo, btw great generator!

englercj commented:
@adamdanielsdesign Thanks, thats what I generally use as a starting point. After my 10th module using the same pattern I made the generator lol.

adamdanielsdesign commented:
@englercj 
**Update:** 
- Implemented the generator
- Modified all the classes
  * Renamed them removing the 'spine.' prefix from the first commit
  * Added appropriate usage of module.exports for all the classes.

**Tested**
- Tested on my local with a spine animation as working

englercj commented on src/SpineRuntime/index.js:
```diff
@@ -0,0 +1,78 @@
+    Float32Array: (typeof(Float32Array) === 'undefined') ? Array : Float32Array,
+    Uint16Array: (typeof(Uint16Array) === 'undefined') ? Array : Uint16Array
+};
+spine.Animation = require('./Animation.js');
```
These can be in the object above, no need to create the object then modify it.

adamdanielsdesign replied:
updated

englercj commented on src/SpineRuntime/Animation.js:
```diff
@@ -0,0 +1,70 @@
+var spine = require('../SpineRuntime') || {};
+spine.Animation = function (name, timelines, duration)
```
This bit isn't necessary, just export the class itself, and directly require the needed classes. This isn't really following the CJS pattern.

For example:

```js
function Animation(name, timelines, duration) {
    // ...
}

module.exports = Animation;

Animation.prototype = {
    // ...
};
```

The index.js file already aggregates these files, no need to do this weird double object export.

adamdanielsdesign replied:
For now, my opinion is to keep the structure the same. 

Many of the classes, are written such that spine.SomeClass/SomeProperty is used all over the place.

The update for this would require substantial sub-structure changes, and so lots of additional testing to ensure changes function properly.

As such the top line of ```var spine = require('../SpineRuntime') || {};``` is necessary for the module type configuration for the repo.

However, I agree, aside from cjs's pattern, it's much more standard to leave out the parent object association.

englercj replied:
Well its more than just pattern, this is a circular dependency. Animation.js requires index.js which requires Animation.js. I don't think it is acceptable to leave it like that, you will find many subtle broken things throughout the code with a circular dep like that.

englercj commented on gulp/util/handleErrors.js:
```diff
@@ -0,0 +1,19 @@
+var gutil = require('gulp-util'),
```
I need to update the generator, I found a better way to handle edge-case errors from browserify. Checkout pixi's `handleErrors.js` file for an example:

https://github.com/GoodBoyDigital/pixi.js/blob/dev/gulp/util/handleErrors.js

adamdanielsdesign replied:
updated

englercj commented on gulp/util/bundle.js:
```diff
@@ -0,0 +1,52 @@
+    return this.bundle()
+        .on('error', handleErrors.handler)
+        .pipe(handleErrors())
+        .pipe(source('pixi-spinejs.js'))
```
spinejs.js? lol, `pixi-spine.js` should be fine :P

adamdanielsdesign replied:
updated

englercj commented on gulp/tasks/test.js:
```diff
@@ -0,0 +1,8 @@
+
+gulp.task('test', function (done) {
+    return gulp.src(paths.tests, { read: false })
+        .pipe(mocha());
```
Are there actually any tests?

adamdanielsdesign replied:
No, I left it in there, later there may be.  
So I left the structure in place.

englercj commented on gulp/tasks/jsdoc.js:
```diff
@@ -0,0 +1,11 @@
+    exec = require('child_process').exec;
+
+gulp.task('jsdoc', function (done) {
+    exec('npm docs', function (err, stdout, stderr) {
```
Is there actually any jsdoc comments?

adamdanielsdesign replied:
Not currently, however probably lots to come! 
So I left it in there.

englercj commented on README.md:
```diff
@@ -1,3 +1,18 @@
-# pixi-spine
-Pixi.js plugin that enables Spine support.
+# pixi-spine.js
```
I'd remove the .js from here.

adamdanielsdesign replied:
updated

englercj commented on LICENSE:
```diff
@@ -0,0 +1,14 @@
+Copyright (c) 2015, Mat Groves
```
Careful, the spine's runtime code (though adapted) is in here. Make sure the Spine license and associated copyrights are also present. LIkely best to put into a `SPINE-LICENSE` file.

adamdanielsdesign replied:
removed LICENSE
re-added SPINE-LICENSE

englercj commented on package.json:
```diff
@@ -1,27 +1,63 @@
 {
-  "name": "pixi-spine",
-  "version": "1.0.1",
+  "name": "pixi-spinejs",
```
pixi-spine!

adamdanielsdesign replied:
updated

englercj commented on package.json:
```diff
@@ -1,27 +1,63 @@
-  "peerDependencies": {
-    "pixi.js": "^3.0.0-rc4"
+  "peerDependencies":{
+  	"pixi.js": "^3.0.0-rc4"
```
indentation seems mixed in this file, do you have editorconfig installed in your editor? :)

adamdanielsdesign replied:
Sorry! Gotta get a plugin for my text editor, I did change the settings manually though and read through the config file.

englercj commented on package.json:
```diff
@@ -1,27 +1,63 @@
+    "watchify": "^2.2.1"
+  },
+  "dependencies": {
+  	"brfs": "~1.4.0"
```
Is brfs used in this lib?

adamdanielsdesign replied:
updated
I left it in there because it was an option for how to configure stuff.

englercj commented on src/Spine/index.js:
```diff
@@ -1,10 +1,7 @@
-    spine = require('./SpineRuntime');
-
+var PIXI = require('pixi.js');
+var spine = require('../SpineRuntime');
```
seems weird for the spine runtime to be in a cap-first folder, when it isn't a class file. I recommend changing the runtime's folder to simply `runtime/`

adamdanielsdesign replied:
Later, when going through and writing docs, SpineRuntime could be a "module" name vs "class" name. Also the main index.js for the repo uses it as a module
```
/**
 * @namespace PIXI.spine
 */
module.exports = {
    Spine:          require('./Spine'),
    SpineRuntime:   require('./SpineRuntime')
};
```
For intuitive reasons, I'd say keep it as is, however I agree that at first it looks like a misnomer, however because the word "...Runtime" is used, it denotes that "Spine" is a class specific to pixi and "SpineRuntime" is the module holding the generic classes from Esoteric.

englercj replied:
I think that is a good argument for keeping it, I'm fine with that.

englercj commented:
Awesome start, added a bunch of recommendations!

adamdanielsdesign commented:
@englercj  Click "show outdated diff" links above; I added some comments to there.

englercj commented:
Looks all good, except the circular dependency I mentioned. I'll let you and @ivanpopelyshev decide how you want to fix that.

ivanpopelyshev commented:
Ok, I see a lot of circular dependencies, I'll try to fix that and test. I'm working with 
https://github.com/adamdanielsdesign/pixi-spine

adamdanielsdesign commented:
@englercj @ivanpopelyshev 
I just tested, yeah with the change to the SpineRuntime/index.js
```
//including the requires in the same object,
var spine = {
  ...
}
spine.Animation = ...

//to 
var spine = {
   ...
  Animation: ...
}
```
It breaks because classes are dependent on those pre-existing values such as "Float32Array"

adamdanielsdesign commented:
@ivanpopelyshev 
You'll need to do a fresh pull to have a working base to start from, or change the SpineRuntime/index.js file to the new one from the recent commit https://github.com/adamdanielsdesign/pixi-spine/commit/18d421fa38357ec659a62a7148d675a2bdd49085

englercj commented:
> I just tested, yeah with the change to the SpineRuntime/index.js
> ...
> It breaks because classes are dependent on those pre-existing values such as "Float32Array"

Only the things you are exporting to the public should be in the index file, the other stuff like typed array stuff should be in a util file internally that all the classes can access.

adamdanielsdesign commented:
**Update:**

Commit : https://github.com/adamdanielsdesign/pixi-spine/commit/2ce40160422ef86db45c1cbae32b4799cfda3c8f

- Put SpineUtil/ as a separate class.
- Refactored the classes to remove the circular dependency issue.
  * included only needed classes
  * method used was to add them on to the spine. object vs doing actual changes to the associated code.

ivanpopelyshev commented:
@adamdanielsdesign @englercj 

[19:35:04] [ERROR] Error: Cannot find module 'pixi.js' from 'd:\pixi-games\adam-pixi-spine\src\Spine'
=> no binary files.

what am I doing wrong?

adamdanielsdesign commented:
- clone pixijs as usual
- replace the pixi-spine module with the test repo:
pixijs/node-modules/pixi-spine/< put the test repo here >

englercj commented:
Or, you could have pixi-spine and pixi.js cloned in two separate folders. Then globally link pixi-spine:

```js
cd pixi-spine && npm link
```

Then link it into pixi.js

```js
cd pixi.js && npm link pixi-spine
```

Npm will create the necessary symlinks for you and now your pixi-spine repo is linked into the node_modules folder of pixi.js.

ivanpopelyshev commented:
2. after I set "curly" : false, "eqeqeq": false, I found that there are errors:
we need to add 

```javascript
var PIXI = require('pixi.js')
```

in Atlas.js

adamdanielsdesign commented:
sorry! left it out.

adamdanielsdesign commented:
@ivanpopelyshev  sorry, probably could've let you make a pull for that one!

ivanpopelyshev commented:
@adamdanielsdesign I dont mind that you are making almost all the work :) At the moment I have several projects and its kind of difficult to multitask.

ivanpopelyshev commented:
Anyway, I will make sure that any person will be able to setup this project ;)

ivanpopelyshev commented:
Ok, I have a build, but it doesnt work in my app like PIXI2:

Uncaught ReferenceError: PIXI is not defined

Why cant I just add "bin/pixi.js" as a script?

adamdanielsdesign commented:
Pixi2 is using straight concat vs browserify. So the require() syntax isn't available there where pixi-spine is as a plugin.

ivanpopelyshev commented:
I mean actual pixi.js ee3 branch. I create a build, add "bin/pixi.js" into my app and it doesnt see "PIXI" variable

adamdanielsdesign commented:
Pixi-Spine isn't a plugin in rc2.

Try the 'dev' branch, it's as a node_module there.

ivanpopelyshev commented:
Ok, now it works, I'm porting my latest project to it.

anyway, trying to build pixijs/dev branch:

W117  Line 25, Col 37  'define' is not defined
W117  Line 27, Col 5   'define' is not defined

adamdanielsdesign commented:
code snippet from those errors?

ivanpopelyshev commented:
@adamdanielsdesign js-hint fails in pixijs dev branch:

```javascript
if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(function () {
        return core;
    });
}
```

ivanpopelyshev commented:
By the way, do you know where did AnimCache go? I found TextureCache in PIXI.utils.TextureCache

adamdanielsdesign commented:
I just did a fresh pull, now I'm getting the same errors.

adamdanielsdesign commented:
W117  Line 25, Col 37  'define' is not defined.
 W117  Line 27, Col 5   'define' is not defined.
W117  Line 87, Col 5  'ths' is not defined.

adamdanielsdesign commented:
it's an error with the pixi repo, not the pixi-spine repo.

ivanpopelyshev commented:
Ok, both mine build and pixi.js dev in "bin" folder have one problem:

PIXI.Spine not found.

adamdanielsdesign commented:
Got:
Error: The spineData param is required.

adamdanielsdesign commented:
Where are you getting "PIXI.Spine not found."

ivanpopelyshev commented:
Ok, i got it, i have to use "PIXI.spine.Spine". 
Now, previously I passed "lalala.json" file in it, but now I have to manually implement AnimCache, right?

adamdanielsdesign commented:
dunno, not sure.

adamdanielsdesign commented:
Where in src is AnimCache?

ivanpopelyshev commented:
Ok, now it works.

```javascript
var AnimCache = {};
assetLoader.load(function (loader, resources) {
    for (var key in resources) if (resources.hasOwnProperty(key)) {
        if (resources[key].spineData)
            AnimCache[key] = resources[key].spineData;
    }
    START_THE_GAME(AnimCache);
}
```

adamdanielsdesign commented:
What is AnimCache I haven't used it?

adamdanielsdesign commented:
Also, ```gulp build --exclude spine``` seems to still include Spine

ivanpopelyshev commented:
How it worked in v2:
1) I pass all json files into AssetLoader.
2) AssetLoader saves spineData's into PIXI.AnimCache[]
3) new PIXI.Spine('lalala.json') takes data from PIXI.AnimCache[]

http://kids.matroid.ru/test/ - IT WORKS!!!

adamdanielsdesign commented:
Do I need to add AnimCache then?
```coffeescript
PIXI.loader
        .add('bot','storybot.json')
        .load(onAssetsLoaded)

#then...
onAssetsLoaded = (loader,res) ->
        app.bot = new PIXI.spine.Spine(res.bot.spineData)
```

adamdanielsdesign commented:
Also, are you still getting the errors with straight gulp?

ivanpopelyshev commented:
In pixi-spine or pixijs?

adamdanielsdesign commented:
pixijs

ivanpopelyshev commented:
It works if I remove that "define" from index.js

adamdanielsdesign commented:
```
//this?
// export for AMD if necessary
if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(function () {
        return core;
    });
}
```

ivanpopelyshev commented:
yep

adamdanielsdesign commented:
Still getting
```src/core/graphics/GraphicsData.js
-------------------------------------------------------------------------------------------------
 W117  Line 87, Col 5  'ths' is not defined.
```

adamdanielsdesign commented:
Should that be 'this'?

adamdanielsdesign commented:
fixed it

ivanpopelyshev commented:
LOL :)

adamdanielsdesign commented:
Getting:
```
resource.data
//instead of 
resource.spineData
```

ivanpopelyshev commented:
"data" is what you get from json file.
"spineData" contains that and everything else that loaded (atlas, texture)

adamdanielsdesign commented:
Only getting data property, no spineData property with:
```coffeescript
        app.AnimCache = {};
        for key of res
            console.log 'key= ', key
            if (res.hasOwnProperty(key))
                console.log res[key]
                if (res[key].spineData)
                    app.AnimCache[key] = res[key].spineData
```

ivanpopelyshev commented:
look at spineAtlasParser

```javascript
// skip if no data, its not json, or it isn't atlas data
        if (!resource.data || !resource.isJson || !resource.data.bones)
        {
            return next();
        }
```

```javascript
        resource.spineData = skeletonData;
        resource.spineAtlas = spineAtlas;
```

englercj commented:
Most likely you didn't update your resource-loader version to 1.4.0. That was when isJson was added to resource objects.

Also thanks for the shout out about `ths`, fixed that :P

adamdanielsdesign commented:
works!

ivanpopelyshev commented:
@englercj My demo works with pixiv3: http://kids.matroid.ru/test/ . 

The only thing I dont understand: why do we need gulp in this module? It wont build without pixijs anyway.

englercj commented:
@ivanpopelyshev As of today that is how it is structured *because* pixi requires it manually right now *because* there is no gulp task here.

Ideally this is a separate module completely, and if someone wants spine they come here and download the pixi-spine built file and include it in addition to pixi. Or if they are using browserify their project has a dependency on pixi.js and pixi-spine. They then just require both.

The only reason pixi.js has a require('pixi-spine') right now is there is no build stuff in here for it to be standalone. Once it is standalone, I am removing it from the main repo.

ivanpopelyshev commented:
Ok, now I understand it. I suggest to merge this PR. Do you agree?

englercj commented:
/agree

adamdanielsdesign commented:
Great!
